Add A New Truffle Random Data Source Example

riccardopersiani opened the pull request:
### What this PR is:

*Add the random-datasource example* to the **truffle-examples**.

***

### To Do: 

__❍__  @gskapka @D-Nice  Please, could you audit and test the code?

***

### Note:

1) I tested it on Rinkeby with: 

`truffle test --network rinkeby`

2) Do you guys have any idea concerning other tests to add in `/test/random-example-test.js`?

D-Nice commented:
@gskapka any feedback on this?

gskapka commented on solidity/truffle-examples/random-datasource/test/random-example-test.js:
```diff
@@ -0,0 +1,35 @@
+  const address = accounts[0]
+  
+  before(async () => (
+	{ contract } = await randomExample.deployed(),
```
Inconsistent spacing.

riccardopersiani replied:
linted

gskapka commented on solidity/truffle-examples/random-datasource/test/random-example-test.js:
```diff
@@ -0,0 +1,35 @@
+ 
+  it('Callback should have logged a new Random Number uint event', async () => {
+    console.log("address: ", address)
+    const {returnValues} = await waitForEvent(events.newRandomNumber_uint)
```
not linted.

gskapka commented on solidity/truffle-examples/random-datasource/test/random-example-test.js:
```diff
@@ -0,0 +1,35 @@
+	{ events } = new web31.eth.Contract(contract._jsonInterface, contract._address) 
+  ))
+ 
+  it('Callback should have logged a new Random Number uint event', async () => {
```
~~uint~~

gskapka commented on solidity/truffle-examples/random-datasource/test/random-example-test.js:
```diff
@@ -0,0 +1,35 @@
+      await methods.update().send({from: address, gas: gasAmt})
+      assert.fail('Update transaction should not have succeeded!')
+    } catch (e) {
+      assert.isTrue(e.message.startsWith(`${expErr}`), `Expected ${expErr} but got ${e.message} instead!`)
```
Does this still work with newer truffle? I had to change it for the latest example update I did IIRC...

gskapka commented on solidity/truffle-examples/random-datasource/test/utils.js:
```diff
@@ -0,0 +1,24 @@
+const fs = require('fs')
+const path = require('path')
+const API_PATH = path.join(__dirname, '../', 'apikeys.js')
```
```suggestion
const API_PATH = path.resolve(__dirname, '../apikeys.js')
```

riccardopersiani replied:
Is resolve a best practice?

gskapka replied:
No, it's just easier here since it requires 2 params instead of three.

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
```diff
@@ -0,0 +1,66 @@
+
+## :page_with_curl:  _Instructions_
+
+Since the random datasource is currently available on the Ethereum mainnet and on all Ethereum public testnets only (Rinkeby, Kovan, Ropsten-revival); and it is *not integrated yet with private blockchains/testrpc/browser-solidity-vmmode*, you will need to **get a mnemonic passphrase and an infura key** to make the example work.
```
...and some test net ethereum!

riccardopersiani replied:
Yeah, for now goerli doesn't

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
```diff
@@ -0,0 +1,66 @@
+**B)** Get a mnemonic passphrase, an easy way is to get the Metamask one:
+[metamask.io](https://metamask.io/)
+
+**C)** Create a new file `apikeys.js` or `.env` and *add the mnemonic passphrase and the infura key to it*, such as:
```
Just adding something to a `.env` file doesn't magically make it an environment variable, and so neither the code I wrote in the `truffle.js` nor that you copy/pasted into utils (they should share it, btw, with proper exporting, rather than copy/pasting) would pick up the keys from a `.env` file if the user chooses this option.

riccardopersiani replied:
This line `require('dotenv').config()` adds the `.env` file variables to the environment variable stack.

gskapka replied:
And where is this line? Because as of your latest commit @ time of writing (2ed7a8a) in this PR, neither this line nor the package in the `package.json` exist.

I am aware that that is how you parse `.env` to achieve the environment variable, but I'm pointing out the lack of it in this repo, making the `.env` pointless, and the recommendation for using it in the `README` also pointless.

riccardopersiani replied:
The line is the fix...

gskapka replied:
I am well aware, but you stated the above line as if:

a) It's already in the repo and working
& b) I don't know how to use an environment variable in a node project.

But the point I'm making isn't me asking how to do it, it's that you've suggested a user (who is likely to be a beginner given that they are looking at our truffle examples) to use an environment variable in a `.env` file when that won't work in the repo as it stands at the time of reviewing. 

And so I've raised this entire point in the review precisely because it's _not_ in the repo, and so _doesn't_ work.

riccardopersiani replied:
No, I stated:

this line will now fix the issue. 

If it was not clear the point I'm making with: 

> This line require('dotenv').config() adds the .env file variables to the environment variable stack.

is: there was an issue you pointed out, and I solved it adding that line. I wasn't claiming for obvious reasons that was already in the repo and working.

gskapka replied:
>This line require('dotenv').config() adds the .env file variables to the environment variable stack.

You statement is declarative and conveys no sense of having made a fix whatsoever, nor intention to do so. It simply points out what we all know about how to use an environment variable in a node project.

This review is me pointing out things that need fixing.

riccardopersiani replied:
I am confused. Again, what I want to say here is that editing the file adding `require('dotenv').config()` fix now the issue you pointed out.

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
````diff
@@ -0,0 +1,66 @@
+```javascript
+// apikeys.js example
+
+module.exports = {
````
Again, you suggest this as a format for both `apikeys` and `.env`, but for the latter this will not work.

riccardopersiani replied:
I was suggesting this just for apikeys

gskapka replied:
In which case where is the bit about how to use the `.env` file for the reader of the tutorial?

riccardopersiani replied:
there is no reference.

gskapka replied:
I know, do you see what I'm getting at...?

riccardopersiani replied:
What's going on here? Do you want me to add more details to the doc?

```
**C)** Create a new file `apikeys.js` or `.env` and *add the mnemonic passphrase and the infura key to it*
```

with the new `require('dotenv').config()` adding variables to `.env` will be enough; but, if it's not so clear, I will add an example like this in the doc:

```
// .env example 

mnemonic = 'word1 word2 ...'
infuraKey = 'f09924890213480...'
```

gskapka replied:
Yes, that's what's going on. If you're mentioning `.env` in the `README`, those `.env` need to work. And currently, as of the reviewed commit 2ed7a8a, they do not.

riccardopersiani replied:
.env details added to the readme

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
````diff
@@ -0,0 +1,66 @@
+}
+```
+
+Please note that those must be kept secure and the one provided here are just to help the user with the tutorial.
````
```suggestion
Please note that these keys must be kept secure. The ones provided here are just placeholders to help with the tutorial.
```

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
```diff
@@ -0,0 +1,66 @@
+
+Then you can proceed with the following console instructions:
+
+**1)** Fire up your favourite console & make sure you have Truffle globally 5 installed:
```
The following instructions all need changing to be brought in line with the other trufle examples, which now use `npx` instead of global installs.

gskapka commented on solidity/truffle-examples/random-datasource/test/utils.js:
```diff
@@ -0,0 +1,24 @@
+
+const fileExists = _path => fs.existsSync(_path)
+
+const getInfuraKey = () => fileExists(API_PATH)
```
Rather than copy/pasting this from `truffle.js`, please export & resuse.

riccardopersiani commented on solidity/truffle-examples/random-datasource/.gitignore:
```diff
@@ -0,0 +1,13 @@
+*.sw*
```
Removed

riccardopersiani commented on solidity/truffle-examples/random-datasource/contracts/RandomExample.sol:
```diff
@@ -0,0 +1,51 @@
+*/
+
+pragma solidity ^0.5.0;
+
```
Approved

riccardopersiani commented on solidity/truffle-examples/random-datasource/contracts/RandomExample.sol:
```diff
@@ -0,0 +1,51 @@
+   This contract uses the random-datasource to securely generate off-chain N random bytes
+*/
+
+pragma solidity ^0.5.0;
```
Approved

riccardopersiani commented on solidity/truffle-examples/random-datasource/contracts/RandomExample.sol:
```diff
@@ -0,0 +1,51 @@
+
+    constructor() public {
+        oraclize_setProof(proofType_Ledger); // sets the Ledger authenticity proof in the constructor
+        update(); // let's ask for N random bytes immediately when the contract is created!
```
Very good point, approved.

riccardopersiani commented on solidity/truffle-examples/random-datasource/contracts/RandomExample.sol:
```diff
@@ -0,0 +1,51 @@
+            
+            emit newRandomNumber_bytes(bytes(_result)); // this is the resulting random number (bytes)
+            
+            // for simplicity of use, let's also convert the random bytes to uint if we need
```
Approved

riccardopersiani commented on solidity/truffle-examples/random-datasource/migrations/1_initial_migration.js:
```diff
@@ -0,0 +1,5 @@
+var Migrations = artifacts.require("./Migrations.sol");
```
Changed to `const`

riccardopersiani commented on solidity/truffle-examples/random-datasource/test/random-example-test.js:
```diff
@@ -0,0 +1,35 @@
+const RINKEBY_WSS = `wss://rinkeby.infura.io/ws/v3/${getInfuraKey()}`
+const provider = new Web3.providers.WebsocketProvider(RINKEBY_WSS)
+const web31 = new Web3(provider)
+
```
linted

riccardopersiani replied:
If you were referring to the double empty line, now has been removed.

riccardopersiani commented on solidity/truffle-examples/random-datasource/test/random-example-test.js:
```diff
@@ -0,0 +1,35 @@
+const randomExample = artifacts.require('./RandomExample.sol')
+const RINKEBY_WSS = `wss://rinkeby.infura.io/ws/v3/${getInfuraKey()}`
+const provider = new Web3.providers.WebsocketProvider(RINKEBY_WSS)
+const web31 = new Web3(provider)
```
Do you mean something like `const web31 = new Web3(new Web3.providers.WebsocketProvider(RINKEBY_WSS))`?

gskapka replied:
Yes for the inlining but it still needs _naming_ something better.

riccardopersiani commented:
npm version - 5.4.0

### Truffle Random DS

* Tested using both `npx` and `pnpx`.
* Tested with both `apikeys.js` and `.env`

Please @gskapka, could you re-review?

riccardopersiani commented:
Added the missing package "bindings", which wasn't required for the `npm` version I was using.

Moreover, I encountered several problems with the latest versions of npm and node, so I added truffle locally. Seems to have solved the issue.

Please, re-re-review @gskapka.

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
```diff
@@ -0,0 +1,71 @@
+**B)** Get a mnemonic passphrase, an easy way is to get the Metamask one:
+[metamask.io](https://metamask.io/)
+
+**C)** In `ethereum-examples/solidity/truffle-examples/random-datasource`, create a new file `apikeys.js` or `.env` and *add the mnemonic passphrase and the infura key to it*, such as:
```
These instructions come _before_ the main instructions, which latter's first command involves cloning this repo. And so it's confusing, since this earlier line assumes the user has _already_ cloned the repo. 

It's probably best to merge to two instruction sets into one.

gskapka commented on solidity/truffle-examples/random-datasource/package.json:
```diff
@@ -0,0 +1,26 @@
+  },
+  "homepage": "https://github.com/n/a#readme",
+  "dependencies": {
+    "bindings": "^1.5.0",
```
If we're adding truffle too, is this still necessary?

riccardopersiani commented:
> simply hangs the terminal for ~30s

@gskapka - sometimes, I have experienced this as well.
Seems like the _execution time_ is relatively long, but here still I got it passing, downloading a brand new repo:

![image](https://user-images.githubusercontent.com/10484858/54940615-f76b5680-4f22-11e9-9766-7103e5638a34.png)

Sometimes, when fails due to timeout exceeded, I still can see the random number printed, which means the query was sent and the callback received correctly.

Do you have any idea on how we can reduce the execution time or increase the timeout limit on Truffle?

gskapka commented:
> increase the timeout limit on Truffle?

You can easily find info on how to do that with mocha.

This is more worrying: 

> I still can see the random number printed

Since it plus your screen shot means there're console logs in the tests. I never got as far looking at the `test.js` in the review earlier, so I hope you haven't snuck some console logging in the repo...

***

Re the other issue: it hangs until timeout in `truffle develop` but if you read what I wrote more closely in my review from this morning, I mentioned that when run from the command line via the `README`s instruction: `npx truffle test --network rinkeby` it:

> hangs the terminal for ~30s, after which the process completes

Have you seen this behaviour? I get the cursor back and the process ends after _nothing_ happening for 30s. Was your screen shot done when running from the develop console or your terminal?

riccardopersiani commented:
After the last changes, I was able to test it several times successfully, and also freshly cloning the repo.

I wasn't able to replicate the specific issue you were mentioning.
The issue I was having with Ropsten testing was due to the web-socket not initialised properly, and resulted in a timeout error. 
While on Rinkeby everything seems to be fine.

Changes:

* Remove bindings from package.json: works as you suggested
* Edit Readme: The two instructions set are now merged
* Edit test: Timeout increased for the most expensive test in term of execution time
* Edit test: New Oraclize query log test added

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
```diff
@@ -0,0 +1,68 @@
+
+## :page_with_curl:  _Instructions_
+
+The random datasource is currently available on the Ethereum mainnet and on some Ethereum public testnets only (Rinkeby, Kovan, and Ropsten Revival); and it is *not integrated yet with private blockchains/testrpc/browser-solidity-vmmode*, so you will need to **get a mnemonic passphrase and an infura key** to make the example work.
```
Please remove the bold & italic formatting. Mixing the two and overusing them both looks bad and makes comrehension harder.

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
```diff
@@ -0,0 +1,68 @@
+The random datasource is currently available on the Ethereum mainnet and on some Ethereum public testnets only (Rinkeby, Kovan, and Ropsten Revival); and it is *not integrated yet with private blockchains/testrpc/browser-solidity-vmmode*, so you will need to **get a mnemonic passphrase and an infura key** to make the example work.
+
+**1)** Go on the infura website to get an infura key: [infura.io](https://infura.io).
+The key is a 32 HEX characters string
```
No new line.

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
```diff
@@ -0,0 +1,68 @@
+**1)** Go on the infura website to get an infura key: [infura.io](https://infura.io).
+The key is a 32 HEX characters string
+
+**2)** Get a mnemonic passphrase, an easy way is to get the Metamask one:
```
~~the~~ a Metamask one

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
```diff
@@ -0,0 +1,68 @@
+The key is a 32 HEX characters string
+
+**2)** Get a mnemonic passphrase, an easy way is to get the Metamask one:
+[metamask.io](https://metamask.io/)
```
No newline.

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
```diff
@@ -0,0 +1,68 @@
+
+__`❍ cd ethereum-examples/solidity/truffle-examples/random-datasource && npm install`__
+
+**5)** In `ethereum-examples/solidity/truffle-examples/random-datasource`, create a new file `apikeys.js` or `.env` and *add the mnemonic passphrase and the infura key to it*, such as:
```
Rm italics.

gskapka commented on solidity/truffle-examples/random-datasource/README.md:
```diff
@@ -0,0 +1,68 @@
+
+## :black_nib: Notes:
+
+__❍___ Note that the Infura key and the Metamask passphrase must be kept secure. The ones provided in the readme are just placeholders to help the user with the tutorial.
```
You've snuck an extra `_` here.

```suggestion
__❍__ Note that your own Infura key and the Metamask passphrase must be kept secure. The ones provided here are simply placeholders.
```

gskapka commented on solidity/truffle-examples/random-datasource/contracts/RandomExample.sol:
```diff
@@ -0,0 +1,73 @@
+            emit newRandomNumber_bytes(bytes(_result)); // emit the resulting random number (in bytes)
+
+            /**
+             * for simplicity of use, let's also convert the random bytes to uint:
```
Sentences w/ capitals please.

gskapka commented on solidity/truffle-examples/random-datasource/contracts/RandomExample.sol:
```diff
@@ -0,0 +1,73 @@
+             * want to get. The variable maxRange should never be greater than 2^(8*N), where N is
+             * the number of random bytes we had asked the datasource to return.
+             * finally, we perform the modulo maxRange of the sha3 hash of the random bytes casted
+             * to uint to obtain a random number ∈ [0, maxRange - 1].
```
Probably best to lose the set-theory notation since it's more likely to confuse than elucidate. 

Also, capitals at the start of sentences.

Also also, ~~casted~~ cast.

gskapka commented on solidity/truffle-examples/random-datasource/test/random-example-test.js:
```diff
@@ -0,0 +1,68 @@
+const {
+    waitForEvent,
+    getInfuraKey
```
Not linted.

gskapka commented on solidity/truffle-examples/random-datasource/test/random-example-test.js:
```diff
@@ -0,0 +1,68 @@
+const Web3 = require('web3')
+const randomExample = artifacts.require('./RandomExample.sol')
+const RINKEBY_WSS = `wss://rinkeby.infura.io/ws/v3/${getInfuraKey()}`
+const web3Socket = new Web3(new Web3.providers.WebsocketProvider(RINKEBY_WSS))
```
Better naming than before!

gskapka commented on solidity/truffle-examples/random-datasource/test/random-example-test.js:
```diff
@@ -0,0 +1,68 @@
+      0,
+      'A random number should have been retrieved from Oraclize call!'
+    )
+  }).timeout(600000)
```
I hadn't seen this way of changing the timeout of a single mocha test - nice!

gskapka commented on solidity/truffle-examples/random-datasource/test/utils.js:
```diff
@@ -0,0 +1,25 @@
+
+const fileExists = _path => fs.existsSync(_path)
+
+const getInfuraKey = () => fileExists(API_PATH)
```
This is still a copy/paste.

riccardopersiani replied:
Changed that function to the more general getExternalVariable, and exported where needed.

riccardopersiani commented:
Approved small changes have been added with a new image for the passing tests.
